packages/core/src/types.ts

```diff
@@ -886,10 +921,20 @@ export enum ActionTypes {
   Choose = 'xstate.choose'
 }
 
-export interface RaiseAction<TEvent extends EventObject> {
+// Used as a "phantom type" for built-in actions
+export interface BuiltInActionObject<TContext, TEvent extends EventObject> {
+  type: string;
+  __context: TContext;
+  __event: TEvent;
+}
+
+export type RaiseAction<TEvent extends EventObject> = BuiltInActionObject<
+  any,
+  TEvent
+> & {
```

Comment:
TS team recommendation is to prefer interfaces (and extending from interfaces, rather creating types that are intersections) - they supposedly perform better (perf-wise) in some situations.

Reply:
I had interfaces at first; type inference didn't work. It magically works with types instead. (Try it out yourself)

Reply:
It seems to work in the same way with interfaces and types: https://github.com/statelyai/xstate/commit/7e1448c5be9ec4f36409d4af39e32083f30435ed

Reply:
🍒 ⛏️